Record the last fatal error message for crash reporting on Windows

hjyamauchi opened the pull request:
AFAICT, `crashreporter` is specific to Darwin and currently it's not possible to record the fatal error message in a crash dump on Windows. Use a similar technique to make it available for Windows crash dump tools.

hjyamauchi committed:
Record the last fatal error message for crash reporting on Windows

hyp commented:
nice! is there a good way to test this change as well?

tristanlabelle commented on stdlib/public/runtime/Errors.cpp:
```diff
@@ -299,6 +303,34 @@ reportOnCrash(uint32_t flags, const char *message)
+#if defined(_WIN32)
+  // The last fatal error message is accessible for crash reporting on
+  // Windows.
+  char *oldMessage = nullptr;
```
We could reuse the same code as above by defining `char** lastFatalErrorMessage = ` `&gCRAnnotations` or `&gLastFatalErrorMessage` based on preprocessor values.

hjyamauchi replied:
Done

tristanlabelle commented on stdlib/public/runtime/CrashReporter.cpp:
```diff
@@ -33,3 +33,10 @@ struct crashreporter_annotations_t gCRAnnotations __attribute__((
 
 #endif
+
+#if defined(_WIN32)
```
Is the redeclaration necessary given `#include "swift/Runtime/Debug.h"`?

hjyamauchi replied:
Yes, note that the one in `Debug.h` is extern and this here is the real definition / storage location.

tristanlabelle reviewed (approved):
Thanks, that'll be very helpful!

hjyamauchi commented:
> nice! is there a good way to test this change as well?

Added a test

hjyamauchi commented:
PTAL

hyp commented on test/Runtime/fatal-message-crash-dump-win.swift:
```diff
@@ -0,0 +1,43 @@
+        }
+        let ppLastFatalErrorMessage = unsafeBitCast(
+            GetProcAddress(hSwiftCore, "gLastFatalErrorMessage"),
+            to: UnsafeMutablePointer<UnsafeMutablePointer<UInt8>>.self)
```
Is it possible to cast it to `Optional<UnsafeMutablePointer<Optional<UnsafeMutablePointer<UInt8>>>>` instead? That way you can compare it for nil using optional instead of the bit pattern.

hjyamauchi replied:
Done, PTAL

hyp reviewed (approved):
Thanks!

DougGregor commented:
@swift-ci please smoke test

hyp commented:
@hjyamauchi Also, could you add a test case for `try!` failure too? I think that error description should now be captured from the same output

compnerd commented:
Also consider: https://github.com/apple/swift/pull/72785 - this is currently Windows specific which seems like a shame. Additionally, I am concerned about the uniqueness of the address of the storage when used across modules on PE/COFF where the address of a global imported would be the IAT slot IIRC.

hyp commented:
@hjyamauchi here's an example of a try failure:

```
enum Errs: Error {
 case getMyDescription(String)
}

func throwings(_ i: Int) throws {
  if i == 0 {
    throw Errs.getMyDescription("hello")
  }
}

func testMe() {
  try!  throwings(0)
}

testMe()
```

hjyamauchi commented:
@hyp Thanks. I added a try! test.

@compnerd Would it make sense (and are you willing) to adapt the tests in this PR into https://github.com/apple/swift/pull/72785 ? (or I could also work on that if preferred)

al45tair commented:
I'm a little conflicted about this. I think @compnerd is right that there's an issue with the global storage on Windows, but on a more general note I'm not sure how users are expected to make use of this (on non-Darwin platforms).

What's really needed (IMO) is a Windows version of the crash handling code that I added for Darwin and Linux, and then a little bit of extra work to better integrate `fatalError()` with that on all platforms. The Darwin crash reporter code will stay, because that's still useful there, but on Linux and Windows getting the extra data into a crash log like this isn't trivial.

From a platform standpoint, on Windows it might actually make sense for `fatalError()` to report things to the Event Log using `ReportEvent()`. They'd end up as separate events in the application event log that you'd find before the actual crash, but you wouldn't need to have a crash handler installed or enabled to get the data.

compnerd commented:
Having a custom stream in the mini dump is another option on windows that we shouldn't rule out imo. That does require that we install a custom dump handler though. I'm hesitant to expose a stream writer as SPI as that would cause additional code execution in the address space after a fatalError.

hjyamauchi commented:
> I'm a little conflicted about this. I think @compnerd is right that there's an issue with the global storage on Windows, but on a more general note I'm not sure how users are expected to make use of this (on non-Darwin platforms).

A way that I'm thinking of using this is through sentry/crashpad as in https://github.com/thebrowsercompany/swift-sentry/pull/35 This works and puts the fatal error messages into sentry crash reports through a minidump.
 
> What's really needed (IMO) is a Windows version of the crash handling code that I added for Darwin and Linux, and then a little bit of extra work to better integrate `fatalError()` with that on all platforms. The Darwin crash reporter code will stay, because that's still useful there, but on Linux and Windows getting the extra data into a crash log like this isn't trivial.

Can you point to "the crash handling code that I added for Darwin and Linux", if available?

> From a platform standpoint, on Windows it might actually make sense for `fatalError()` to report things to the Event Log using `ReportEvent()`. They'd end up as separate events in the application event log that you'd find before the actual crash, but you wouldn't need to have a crash handler installed or enabled to get the data.

Do you mind clarifying "report things to the Event Log using `ReportEvent()`"? I'm not sure what `ReportEvent` refers to.

compnerd commented:
> A way that I'm thinking of using this is through sentry/crashpad as in https://github.com/thebrowsercompany/swift-sentry/pull/35 This works and puts the fatal error messages into sentry crash reports through a minidump.

That is a product level decision, not a language level. The only two "blessed" error reporting mechanisms for Windows would be the custom minidump writer and `ReportEvent`. The latter being preferable as the user may wish to have control over the minidump emission and us squatting that space would be a problem. Sentry and crashpad are both outside the scope of supportable tools for the language runtime. 

> Do you mind clarifying "report things to the Event Log using ReportEvent()"? I'm not sure what ReportEvent refers to.

https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-reporteventw

al45tair commented:
> Can you point to "the crash handling code that I added for Darwin and Linux", if available?

The Linux crash handler is here: [CrashHandlerLinux.cpp](https://github.com/apple/swift/blob/main/stdlib/public/runtime/CrashHandlerLinux.cpp). There's a similar one for macOS, and I've been meaning to get around to writing something for Windows too. It's on by default for Linux, and I was intending to have it on by default for Windows also (subject to discussing it with @compnerd, since it does mean running code in a crashed process, which is all kinds of fun).

al45tair commented:
> The latter being preferable as the user may wish to have control over the minidump emission and us squatting that space would be a problem.

Agreed. The existing Linux and macOS crash handlers both make sure to crash the process the same way it would have crashed without them.

hjyamauchi commented:
Closing this in favor of https://github.com/apple/swift/pull/72785
